tenpy/models/hofstadter.py

```diff
@@ -92,50 +173,25 @@ def init_lattice(self, model_params):
         bc_y = 'periodic' if bc_y == 'cylinder' else 'open'
         if bc_MPS == 'infinite' and bc_x == 'open':
             raise ValueError("You need to use 'periodic' `bc_x` for infinite systems!")
-        lat = Square(Lx, Ly, site, order=order, bc=[bc_x, bc_y], bc_MPS=bc_MPS)  # LS are Lx, Ly, site defined?
+        lat = Square(Lx, Ly, site, order=order, bc=[bc_x, bc_y], bc_MPS=bc_MPS)
         return lat
 
     def init_terms(self, model_params):
         Lx = self.lat.shape[0]
         Ly = self.lat.shape[1]
+        mx = get_parameter(model_params, 'mx', 0, self.name)
+        my = get_parameter(model_params, 'my', 0, self.name)
         Jx = get_parameter(model_params, 'Jx', 1., self.name)
         Jy = get_parameter(model_params, 'Jy', 1., self.name)
         phi_pq = get_parameter(model_params, 'phi', (1, 3), self.name)
         phi = 2 * np.pi * phi_pq[0] / phi_pq[1]
         phi_ext = get_parameter(model_params, 'phi_ext', 0., self.name)
         mu = get_parameter(model_params, 'mu', 1., self.name, True)
         gauge = get_parameter(model_params, 'gauge', 'landau_x', self.name)
+        hop_x, hop_y = gauge_hopping(gauge, Lx, Ly, mx, my, Jx, Jy, phi, phi_pq)
 
         # 6) add terms of the Hamiltonian
         self.add_onsite(-mu, 0, 'N')
-
-        if gauge == 'landau_x':
-            assert Lx % phi_pq[1] == 0, "Flux density inconsistent with Lx in Landau-x gauge."
```

Comment:
Note all assertions of this form were deleted as add_coupling() is expected to throw an error when the hopping amplitude array is incommensurate with the lattice dimensions, and magnetic unit cell dimensions are now determined based on the flux density.

Reply:
If one just tries to initialize a model with e.g. `Lx=1`, one gets an error like
`ValueError: incomensurate len for tiling from 3 to 1`, with a not a full traceback like
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/scratch/ga93non/tmp/tenpy_forks/jhauschild/tenpy/models/hofstadter.py", line 163, in __init__
    CouplingMPOModel.__init__(self, model_params)
  File "/scratch/ga93non/tmp/tenpy_forks/jhauschild/tenpy/models/model.py", line 1171, in __init__
    self.init_terms(model_params)
  File "/scratch/ga93non/tmp/tenpy_forks/jhauschild/tenpy/models/hofstadter.py", line 208, in init_terms
    hop_y = self.coupling_strength_add_ext_flux(hop_y, dy, [0, phi_ext])
  File "/scratch/ga93non/tmp/tenpy_forks/jhauschild/tenpy/models/model.py", line 948, in coupling_strength_add_ext_flux
    strength = to_array(strength, c_shape)
  File "/scratch/ga93non/tmp/tenpy_forks/jhauschild/tenpy/tools/misc.py", line 64, in to_array
    a.shape[i], shape[i]))
ValueError: incomensurate len for tiling from 3 to 1
```

If you think this is readable/clear enough, I'm happy with it.

Reply:
Yes, that looks absolutely fine!